Add streaming speech enhancemnt inference.

LiChenda opened the pull request:
Hi,
I added `espnet2/bin/enh_inference_streaming.py` to support the streaming inference of some causal enh-models.

Now, the separators can implement a `forward_stream` method (a new abstract method in Enh Separator), and then it can be called in `espnet2/bin/enh_inference_streaming.py`.
Currently, I implement the `forward_stream` function for three separators:

1.  TCN (Consistency is OK, performance needs to be optimized)
2.  LSTM (Consistency is OK)
3.  SkiM (Consistency is OK, currently only  support `no segment padding` mode, `segment padding` will be updated soon)

Some separated waves (upper is from old sequence inference, lower is from the streaming inference):

<img width="600" alt="image" src="https://user-images.githubusercontent.com/9349732/226893263-ea58ea76-71cc-4c06-be87-bed350a91fb0.png">

Future works:

- [x] Prepare a streaming demo in Colab notebook
- [ ] Performance optimization and support more separators.

------

Updated: 

I prepared a streaming demo for speech separation. The notebook demo requires the new features of this PR, so I'll update it to the [demo repo](https://github.com/espnet/notebook) when this PR gets merged.

Now, anyone with the following link can post comments on it, it would be great if you could leave your comments and help improve it.:
https://colab.research.google.com/drive/17vd1V78eJpp3PHBnbFE5aVY5uMxQFL6o?usp=sharing

LiChenda commented:
Hi, I also prepared a streaming demo for speech separation. The notebook demo requires the new features of this PR, so I'll update it to the [demo repo](https://github.com/espnet/notebook) when this PR gets merged.

Now, anyone with the following link can post comments on it, it would be great if you could leave your comments and help improve it.:
https://colab.research.google.com/drive/17vd1V78eJpp3PHBnbFE5aVY5uMxQFL6o?usp=sharing

sw005320 commented:
Cool!
Would it be better to include the installation part?

<img width="970" alt="image" src="https://user-images.githubusercontent.com/11741550/228283373-059498e9-30cc-4ef4-81a9-cf1463fd8184.png">

LiChenda commented:
> Cool! Would it be better to include the installation part?
> 
> <img alt="image" width="970" src="https://user-images.githubusercontent.com/11741550/228283373-059498e9-30cc-4ef4-81a9-cf1463fd8184.png">

Done! Now you @sw005320  can run it in the Colab notebook. But the RTF is about 2 with the Colab's runtime... In my local machine, it's smaller than 1 😂. Seems we need more optimization.

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+            self.streaming_states,
+            _,
+        ) = self.enh_model.separator.forward_streaming(
+            frame_feature, self.streaming_states
```
Better to also add `f_lens` as an argument as some models may require such information.

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+        lengths = to_device(lengths, device=self.device)
+
+        # b. Enhancement/Separation Forward
+        frame_feature, f_lens = self.enh_model.encoder(speech_mix, lengths)
```
Do we need to consider possible buffering in the encoder and decoder as well if we feed the frames one-by-one? I mean if the encoder/decoder is also a multi-layer module, it might be wrong to simply use the default forward function.

Emrys365 replied:
Certain cases may make this problem critical. For example, if an STFT encoder with `center=True` is used, it will apply zero-paddings to both size and the output will be multiple frames.

Emrys365 replied:
In cases where a multi-layer encoder/decoder is used, it might be difficult to define an exact `frame_size` or `hop_size`.

LiChenda replied:
Add `forward_streaming` for the encoder and decoder now. 

> In cases where a multi-layer encoder/decoder is used, it might be difficult to define an exact frame_size or hop_size.
It can be determined by the ABSEncoder and ABSEncoder abstract function:
```
  @property
  def frame_size(self) -> int:
      raise NotImplementedError

```

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+
+        lengths = torch.cat([lengths, lengths])
+
+        waves = self.enh_model.decoder(frame_out, lengths)[0].unbind(dim=0)
```
ditto

LiChenda replied:
Done.

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+        speech = batch["speech_mix"]
+
+        # split continuous speech into small chunks to simulate streaming
+        speech_sim_chunks, rest_pad = split_audio(speech, frame_size, hop_size)
```
We need to be careful when manually generating the frames because it should be consistent with the procedure defined in the encoder. For example, depending on the padding mode of the encoder, we may need to apply some paddings (can be non-zero) to the left/right side of the original input signal.

Emrys365 replied:
If we do that manually, then we should ignore the builtin padding operations in the actually encoder module. Things become too complicated...

LiChenda replied:
Still, keep it manually outside the model.  Builtin padding is ignored by new `forward_streaming` in the encoder and decoder.

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+            frame_feature, self.streaming_states
+        )
+
+        frame_out = torch.cat(frame_out, 0)
```
Could you annotate the shape of the output?

LiChenda replied:
Done.

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+
+        frame_out = torch.cat(frame_out, 0)
+
+        lengths = torch.cat([lengths, lengths])
```
I cannot understand why **two** `lengths` are concatenated here.

LiChenda replied:
Removed.

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+
+def recursive_dict_update(dict_org, dict_patch, verbose=False, log_prefix=""):
+    """Update `dict_org` with `dict_patch` in-place recursively."""
+    for key, value in dict_patch.items():
+        if key not in dict_org:
+            if verbose:
+                logging.info(
+                    "Overwriting config: [{}{}]: None -> {}".format(
+                        log_prefix, key, value
+                    )
+                )
+            dict_org[key] = value
+        elif isinstance(value, dict):
+            recursive_dict_update(
+                dict_org[key], value, verbose=verbose, log_prefix=f"{key}."
+            )
+        else:
+            if verbose and dict_org[key] != value:
+                logging.info(
+                    "Overwriting config: [{}{}]: {} -> {}".format(
+                        log_prefix, key, dict_org[key], value
+                    )
+                )
+            dict_org[key] = value
+
+
+def build_model_from_args_and_file(task, args, model_file, device):
+    model = task.build_model(args)
+    if not isinstance(model, AbsESPnetModel):
+        raise RuntimeError(
+            f"model must inherit {AbsESPnetModel.__name__}, but got {type(model)}"
+        )
+    model.to(device)
+    if model_file is not None:
+        if device == "cuda":
+            # NOTE(kamo): "cuda" for torch.load always indicates cuda:0
+            #   in PyTorch<=1.4
+            device = f"cuda:{torch.cuda.current_device()}"
+        model.load_state_dict(torch.load(model_file, map_location=device))
+    return model
```
You could simply import these functions from `espnet2.bin.enh_inference` to reduce the number of lines.

LiChenda replied:
Done.

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+    """SeparateSpeech class
+
+    Examples:
+        >>> import soundfile
+        >>> separate_speech = SeparateSpeech("enh_config.yml", "enh.pth")
+        >>> audio, rate = soundfile.read("speech.wav")
+        >>> separate_speech(audio)
+        [separated_audio1, separated_audio2, ...]
```
The examples should be updated.

LiChenda replied:
Done.

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+
+    rest = frame_size - (hop_size + audio_len % frame_size) % frame_size
+
+    if rest > 0:
+        pad = torch.zeros((batch_size, rest), dtype=audio.dtype).to(audio.device)
+        audio = torch.cat([audio, pad], 1)
+
+    pad_aux = torch.zeros((batch_size, hop_size), dtype=audio.dtype).to(audio.device)
```
```suggestion
    if rest > 0:
        pad = torch.zeros((batch_size, rest), dtype=audio.dtype, device=audio.device)
        audio = torch.cat([audio, pad], 1)

    pad_aux = torch.zeros(
        (batch_size, hop_size), dtype=audio.dtype, device=audio.device
    )
```

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,545 @@
+        model_file: Union[Path, str] = None,
+        inference_config: Union[Path, str] = None,
+        ref_channel: Optional[int] = None,
+        normalize_output_wav: bool = False,
```
This argument seems unused.

LiChenda replied:
Add  to unit tests.

Emrys365 commented on espnet2/enh/layers/skim.py:
```diff
@@ -295,19 +328,86 @@ def _padfeature(self, input):
+
+        B, _, N = input_frame.shape
+
+        def empty_seg_states():
+            return (
+                torch.zeros(
+                    1,
+                    B,
+                    self.hidden_size,
+                    device=input_frame.device,
+                    dtype=input_frame.dtype,
+                ),
+                torch.zeros(
+                    1,
+                    B,
+                    self.hidden_size,
+                    device=input_frame.device,
+                    dtype=input_frame.dtype,
+                ),
+            )
```
```suggestion
        def empty_seg_states():
            shp = (1, B, self.hidden_size)
            return (
                torch.zeros(*shp, device=input_frame.device, dtype=input_frame.dtype),
                torch.zeros(*shp, device=input_frame.device, dtype=input_frame.dtype),
            )
```

Emrys365 commented on espnet2/enh/separator/skim_separator.py:
```diff
@@ -129,6 +129,82 @@ def forward(
+    )
+
+    input_feature = torch.randn((1, SEQ_LEN, 128))
+    ilens = torch.LongTensor([SEQ_LEN, SEQ_LEN, SEQ_LEN])
```
```suggestion
    ilens = torch.LongTensor([SEQ_LEN])
```

Emrys365 reviewed (commented):
Could you move some of the tests to `test/espnet2/enh`?

Emrys365 commented on espnet2/enh/encoder/stft_encoder.py:
```diff
@@ -54,3 +66,41 @@ def forward(self, input: torch.Tensor, ilens: torch.Tensor):
+
+        windowed = self._apply_window_func(input)
+
+        feature = torch.fft.fft(windowed)
```
We should condition the operation on the `one_sided` argument:
```python
feature = torch.fft.rfft(windowed) if self.onesided else torch.fft.fft(windowed)
```

Emrys365 commented on espnet2/enh/encoder/stft_encoder.py:
```diff
@@ -54,3 +66,41 @@ def forward(self, input: torch.Tensor, ilens: torch.Tensor):
+
+        windowed = self._apply_window_func(input)
+
+        feature = torch.fft.fft(windowed)
+        if self.stft.onesided:
+            feature = feature[..., 0 : -(self.n_fft) // 2 + 1]
```
We can condition the operation on the `one_sided` argument:
```python
feature = torch.fft.rfft(windowed) if self.stft.onesided else torch.fft.fft(windowed)
```

Emrys365 commented on test/espnet2/enh/decoder/test_stft_decoder.py:
```diff
@@ -58,3 +60,24 @@ def test_STFTDecoder_invalid_type(
+
+def test_conv_enc_dec_streaming():
+    input_audio = torch.randn((1, 16000))
+    ilens = torch.LongTensor(
+        [
+            16000,
+        ]
+    )
```
```suggestion
    ilens = torch.LongTensor([16000])
```

LiChenda commented on espnet2/enh/encoder/stft_encoder.py:
```diff
@@ -54,3 +66,41 @@ def forward(self, input: torch.Tensor, ilens: torch.Tensor):
+            feature = feature[..., 0 : -(self.n_fft) // 2 + 1]
+        feature = feature.unsqueeze(1)
+
+        feature = ComplexTensor(feature.real, feature.imag)
```
Just keep the output type consistent with forward.

LiChenda commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,509 @@
+    return output
+
+
+def split_audio(audio, frame_size, hop_size):
+    # audio: B, T
+    batch_size, audio_len = audio.shape
+
+    # rest = frame_size - (hop_size + audio_len % frame_size) % frame_size
+
+    rest = hop_size - (audio_len - frame_size) % hop_size
+
+    if rest > 0:
+        pad = torch.zeros((batch_size, rest), dtype=audio.dtype, device=audio.device)
+        audio = torch.cat([audio, pad], 1)
+
+    pad_len = math.ceil((frame_size // hop_size) // 2) * hop_size
+
+    pad_aux = torch.zeros((batch_size, pad_len), dtype=audio.dtype, device=audio.device)
+
+    audio = torch.cat([pad_aux, audio, pad_aux], 1)
+
+    new_len = audio.shape[1]
+
+    audio = [
+        audio[:, i * hop_size : i * hop_size + frame_size]
+        for i in range((new_len - frame_size) // hop_size + 1)
+    ]
+
+    return audio, rest
```
I tried your frame segment implementation, and although the number of frames is the same as the `torch.stft`, the spectrum output is not consistent. Would you like to help debug it if you have time?

Emrys365 replied:
OK. I will look at it later today.

Emrys365 replied:
> I tried your frame segment implementation, and although the number of frames is the same as the `torch.stft`, the spectrum output is not consistent. Would you like to help debug it if you have time?

What do you mean by `the spectrum output is not consistent`? I actually obtain the same framed time-domain signals after dropping the last frame from your code:
```python
import math

import librosa
import torch
import torch.nn.functional as F

from espnet2.bin.enh_inference_streaming import split_audio


def frame(x, frame_length, hop_length, dim=-1):
    if x.size(dim) < frame_length:
        raise ValueError(
            "Input is too short (n={:d})"
            " for frame_length={:d}".format(x.size(dim), frame_length)
        )

    if hop_length < 1:
        raise ValueError("Invalid hop_length: {:d}".format(hop_length))

    n_frames = 1 + (x.size(dim) - frame_length) // hop_length
    strides = list(x.stride())

    if dim == -1 or dim == x.dim() - 1:
        shape = list(x.shape[:-1]) + [frame_length, n_frames]
        strides = strides + [hop_length]

    elif dim == 0:
        shape = [n_frames, frame_length] + list(x.shape)[1:]
        strides = [hop_length * strides[0]] + strides

    else:
        raise ValueError("Frame dim={} must be either 0 or -1".format(dim))

    return x.as_strided(shape, strides, storage_offset=0)


if __name__ == "__main__":
    x = torch.rand(2, 16000)
    frame_length = 512
    hop_length = 128
    center = True

    if center:
        signal_dim = x.dim()
        extended_shape = [1] * (3 - signal_dim) + list(x.size())
        pad = int(frame_length // 2)
        x2 = torch.nn.functional.pad(x.view(extended_shape), [pad, pad])
        x2 = x2.view(x2.shape[-signal_dim:])

    # (2, 512, 126)
    splited_th = frame(x2, frame_length, hop_length, dim=-1)
    splited_np = librosa.util.frame(x2.numpy(), frame_length, hop_length, axis=-1)
    # (2, 512, 127)
    splited, rest = split_audio(x, frame_size=frame_length, hop_size=hop_length) 
    splited = torch.stack(splited, dim=-1)

    torch.testing.assert_allclose(splited_th, splited_np)
    torch.testing.assert_allclose(splited_th, splited[..., :-1])

```

Emrys365 replied:
And we can recover the waveform from the framed signal via overlap-add:
```python
length = 16000
start = n_fft // 2 if center else 0
end = -n_fft // 2 if length is None else start + length
channel, n_fft, n_frame = splited.shape

y = torch.nn.functional.fold(
    splited,
    (1, (n_frame - 1) * hop_length + frame_length),
    (1, frame_length),
    stride=(1, hop_length),
).squeeze(2).squeeze(1)[:, start:end]
# in practice, you need to replace `torch.ones_like(splited)` with the actual window function
envelop = torch.nn.functional.fold(
    torch.ones_like(splited),
    (1, (n_frame - 1) * hop_length + frame_length),
    (1, frame_length),
    stride=(1, hop_length),
).squeeze(2).squeeze(1)[:, start:end]
y = y / envelop

torch.testing.assert_allclose(x, y)
```

Emrys365 replied:
In the above, I assume `n_fft == frame_length`. If not, please stick to the value of `n_fft` instead of `frame_length`.

sw005320 commented:
It looks good to me. I'll merge it after the CI is passed.
In addition to the boundary issue in STFT, I want you to consider the following items:
- examples, models, and results for several recipes
- documentations
- making a simple notebook demo

Emrys365 commented on espnet2/enh/separator/tcn_separator.py:
```diff
@@ -114,6 +114,79 @@ def forward(
+        stack=3,
+        bottleneck_dim=32,
+        hidden_dim=64,
+        kernel=3,
+        causal=True,
+        norm_type="cLN",
+    )
+
+    separator.eval()
+
+    print(
+        f"Number of parameters: "
+        f"{humanfriendly.format_size(sum(p.numel() for p in separator.parameters()))}"
+    )
+
+    input_feature = torch.randn((3, SEQ_LEN, 128))
+    ilens = torch.LongTensor([SEQ_LEN, SEQ_LEN, SEQ_LEN])
+
+    with torch.no_grad():
+        start = time.time()
+        seq_output, _, _ = separator.forward(input_feature, ilens=ilens)
+        end = time.time()
+        print(f"seqeunce processing, time cost: {end - start}")
+
+        start = time.time()
+        state = None
+        stream_outputs = []
+        for i in range(SEQ_LEN):
+            frame = input_feature[:, i : i + 1, :]
+            frame_out, state, _ = separator.forward_streaming(frame, state)
+            stream_outputs.append(frame_out)
+        end = time.time()
+        print(f"streaming processing, time cost: {end - start}")
+        for i in range(SEQ_LEN):
+            for s in range(num_spk):
+                torch.testing.assert_allclose(
+                    stream_outputs[i][s], seq_output[s][:, i : i + 1, :]
+                )
+
+    print("Streaming OKey")
```
We can remove this part as the unit test already covers it.

LiChenda replied:
Removed.

Emrys365 commented on espnet2/enh/separator/skim_separator.py:
```diff
@@ -129,6 +129,84 @@ def forward(
+        causal=True,
+        num_spk=2,
+        layer=6,
+        unit=256,
+        segment_size=100,
+    )
+    separator.eval()
+
+    print(
+        "Number of parameters: "
+        f"{humanfriendly.format_size(sum(p.numel() for p in separator.parameters()))}"
+    )
+
+    input_feature = ComplexTensor(
+        torch.randn((1, SEQ_LEN, 128)), torch.randn((1, SEQ_LEN, 128))
+    )
+    ilens = torch.LongTensor([SEQ_LEN])
+
+    with torch.no_grad():
+        start = time.time()
+        seq_output, _, _ = separator.forward(input_feature, ilens=ilens)
+        end = time.time()
+        print(f"seqeunce processing, time cost: {end - start}")
+
+        start = time.time()
+        state = None
+        stream_outputs = []
+        for i in range(SEQ_LEN):
+            frame = input_feature[:, i : i + 1, :]
+            frame_out, state, _ = separator.forward_streaming(frame, state)
+            stream_outputs.append(frame_out)
+        end = time.time()
+        print(f"streaming processing, time cost: {end - start}")
+        for i in range(SEQ_LEN):
+            for s in range(num_spk):
+                torch.testing.assert_allclose(
+                    stream_outputs[i][s], seq_output[s][:, i : i + 1, :]
+                )
+
+    print("Streaming OKey")
```
ditto

Emrys365 commented on espnet2/enh/separator/rnn_separator.py:
```diff
@@ -122,3 +122,86 @@ def forward(
+    num_spk = 2
+    separator = RNNSeparator(
+        input_dim=128,
+        rnn_type="lstm",
+        num_spk=2,
+        layer=7,
+        unit=400,
+    )
+    separator.eval()
+
+    print(
+        "Number of parameters: "
+        f"{humanfriendly.format_size(sum(p.numel() for p in separator.parameters()))}"
+    )
+
+    input_feature = torch.randn((16, SEQ_LEN, 128))
+    ilens = torch.LongTensor([SEQ_LEN] * 16)
+
+    with torch.no_grad():
+        start = time.time()
+        seq_output, _, _ = separator.forward(input_feature, ilens=ilens)
+        end = time.time()
+        print(f"seqeunce processing, time cost: {end - start}")
+
+        start = time.time()
+        state = None
+        stream_outputs = []
+        for i in range(SEQ_LEN):
+            frame = input_feature[:, i : i + 1, :]
+            frame_out, state, _ = separator.forward_streaming(frame, state)
+            stream_outputs.append(frame_out)
+        end = time.time()
+        print(f"streaming processing, time cost: {end - start}")
+        for i in range(SEQ_LEN):
+            for s in range(num_spk):
+                torch.testing.assert_allclose(
+                    stream_outputs[i][s], seq_output[s][:, i : i + 1, :]
+                )
+
+    print("Streaming OKey")
```
ditto

Emrys365 reviewed (commented):
I have some additional comments.

nateanl commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,509 @@
+        >>>
+        >>> for chunk in speech_sim_chunks:
+        >>>     output = separate_speech(chunk)
+        >>>     for channel in range(separate_speech.num_spk):
+        >>>         output_chunks[channel].append(output[channel])
```
```suggestion
        >>>     for spk in range(separate_speech.num_spk):
        >>>         output_chunks[spk].append(output[spk])
```

nateanl commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,509 @@
+    )
+
+    for i, chunk in enumerate(chunks):
+        output[:, i * hop_size : i * hop_size + frame_size] += chunk
```
how's the energy of the overlapped region between two consecutive chunks? I feel the energy is doubled.

Emrys365 replied:
I agree. The current code only works for some specific windows that always sums to 1 in the overlapped regions.

Please refer to the implementation in torchaudio: https://github.com/pytorch/audio/blob/release/0.5/torchaudio/functional.py#L185-L217

LiChenda replied:
Done! And now, all the consistency checks are passed.

nateanl commented on espnet2/enh/decoder/stft_decoder.py:
```diff
@@ -59,3 +73,60 @@ def forward(self, input: ComplexTensor, ilens: torch.Tensor):
+        n_pad_left = (self.n_fft - window.shape[0]) // 2
+        n_pad_right = self.n_fft - window.shape[0] - n_pad_left
+        window = torch.cat(
+            [torch.zeros(n_pad_left), window, torch.zeros(n_pad_right)], 0
```
why the window tensor is zero-padded?

Emrys365 replied:
I think it is for the case where the window size is smaller than `n_fft`. Same as in https://github.com/librosa/librosa/blob/main/librosa/core/spectrum.py#L509

Emrys365 commented on espnet2/enh/decoder/stft_decoder.py:
```diff
@@ -59,3 +73,60 @@ def forward(self, input: ComplexTensor, ilens: torch.Tensor):
 
         return wav, wav_lens
+
+    def _get_window_func(self):
+        window_func = getattr(torch, f"{self.window}_window")
+        window = window_func(self.win_length)
+        n_pad_left = (self.n_fft - window.shape[0]) // 2
+        n_pad_right = self.n_fft - window.shape[0] - n_pad_left
+        window = torch.cat(
+            [torch.zeros(n_pad_left), window, torch.zeros(n_pad_right)], 0
+        )
+        return window
```
This function seems unused in `forward_streaming`, which I think should be used.

LiChenda replied:
solved.

Emrys365 commented on test/espnet2/enh/decoder/test_stft_decoder.py:
```diff
@@ -58,3 +60,24 @@ def test_STFTDecoder_invalid_type(
         y, ilens = decoder(real, x_lens)
+
+
+def test_conv_enc_dec_streaming():
```
Now I think we need to test the consistency for STFTEncoder and STFTDecoder both separately and jointly.
That is, make sure:
  * the streaming mode of STFTEncoder/STFTDecoder is equivalent to the corresponding normal mode
  * the cascaded STFTEncoder and STFTDecoder can reconstruct the input waveform

LiChenda replied:
Done.

LiChenda commented on espnet2/enh/encoder/conv_encoder.py:
```diff
@@ -45,3 +53,7 @@ def forward(self, input: torch.Tensor, ilens: torch.Tensor):
 
         return feature, flens
+
+    def forward_streaming(self, input: torch.Tensor):
```
Unit test added. The default ConvEnc/ConvDec always uses `padding=0`, and it is not configurable at the module level. So we don't need to consider it as all the existing models use `padding=0`.

LiChenda commented:
Thanks for @Emrys365 @nateanl 's comments. I refactored the code slightly, the frame split and merge functions are now moved into `AbsEncoder` and `AbsDecoder`, and implemented by their subclass. That is because the Conv- and STFT-based Enc/Dec act differently.  And that should be more flexible to implement streaming functions for other new Enc/Decs. 

BTW, all the consistency checks have been passed now.

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,458 @@
+        # the main loop for streaming processing
+        for chunk in speech_sim_chunks:
+            # process a single chunk
+            output = separate_speech(chunk)
```
```suggestion
            output = separate_speech(chunk, fs=fs)
```

Emrys365 commented on espnet2/bin/enh_inference_streaming.py:
```diff
@@ -0,0 +1,458 @@
+            (w / abs(w).max(dim=1, keepdim=True)[0] * 0.9).cpu().numpy() for w in waves
+        ]
+
+        # waves = [w.cpu().numpy() for w in waves]
```
remove this line

LiChenda replied:
Done

Emrys365 commented on espnet2/enh/decoder/stft_decoder.py:
```diff
@@ -72,3 +81,116 @@ def forward(self, input: ComplexTensor, ilens: torch.Tensor):
+
+        return output_wav * self._get_window_func()
+
+        # return output_wav
```
ditto

LiChenda replied:
Done

Emrys365 commented on espnet2/enh/decoder/stft_decoder.py:
```diff
@@ -72,3 +81,116 @@ def forward(self, input: ComplexTensor, ilens: torch.Tensor):
+    def _get_window_func(self):
+        window_func = getattr(torch, f"{self.window}_window")
+        window = window_func(self.win_length)
+        n_pad_left = (self.n_fft - window.shape[0]) // 2
+        n_pad_right = self.n_fft - window.shape[0] - n_pad_left
+        # window = torch.cat(
+        #     [torch.zeros(n_pad_left), window, torch.zeros(n_pad_right)], 0
+        # )
```
ditto

LiChenda replied:
Done

Emrys365 reviewed (approved):
Thanks! I only have some minor comments.

sw005320 commented:
Is it OK to merge it?

Emrys365 commented:
Yes, I think it is in a good shape now.
